### PR TITLE
 UPSTREAM: 76663: Fix preemption race conditions on heavy utilized nodes in e2e tests

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
~~The theory I have currently is based on race conditions that could happen when the node gets created and the usage of node objects within the test. I am trying to compute the exact usage depending on the pods before creating pods with other priority classes, so that we won't see resource crunch on the node while creating new pods.~~

EDIT: https://github.com/openshift/origin/pull/22584#issuecomment-483818273

### CODE REVIEW

**Comment by ravisantoshgudimetla** [comment]
To be clear, my understanding after looking at the flakes:

- We are creating pods on nodes which are already heavily utilized but the node.Spec.Allocatable is different from the actual utilizatioin.
- That's why we see that pods are getting created onto nodes but they're actually preempted by pods with higher priority after creation since there is a resource crunch. When, we try to query the pods, since it was already preempted by higher priority pod, we don't see it anymore there by getting 404.

File: vendor/k8s.io/kubernetes/test/e2e/scheduling/preemption.go [hunk 1]
```diff
@@ -133,13 +138,18 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
...
 		pods := make([]*v1.Pod, len(nodeList.Items))
+		allPods, err := cs.CoreV1().Pods(metav1.NamespaceAll).List(metav1.ListOptions{})
 		for i, node := range nodeList.Items {
+			currentCpuUsage, currentMemUsage := getCurrentPodUsageOnTheNode(node.Name, allPods.Items, podRequestedResource)
+			framework.Logf("Current cpu usage and memory usage is %v, %v", currentCpuUsage, currentMemUsage)
 			cpuAllocatable, found := node.Status.Allocatable["cpu"]
 			Expect(found).To(Equal(true))
-			milliCPU := cpuAllocatable.MilliValue() * 40 / 100
+			milliCPU := cpuAllocatable.MilliValue()
+			milliCPU = milliCPU - currentCpuUsage
```
Comment (ravisantoshgudimetla) [inline vendor/k8s.io/kubernetes/test/e2e/scheduling/preemption.go]: Weird that this is returning some -ve value.

**Comment by sjenning** [comment]
`error: failed get version from remote KMS provider` unit flake
/test unit

**Comment by ravisantoshgudimetla** [comment]
The preemption tests are passing now. Re-testing serial tests again.

**Comment by ravisantoshgudimetla** [comment]
Testing again..

**Comment by ravisantoshgudimetla** [comment]
It was a flake  related to image not being available.

**Comment by ravisantoshgudimetla** [comment]
/test e2e-aws-serial

flake related to infrastructure provisioning

File: vendor/k8s.io/kubernetes/test/e2e/scheduling/preemption.go [hunk 2]
```diff
@@ -73,13 +73,20 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
 		// Create one pod per node that uses a lot of the node's resources.
 		By("Create pods that use 60% of node resources.")
 		pods := make([]*v1.Pod, len(nodeList.Items))
+		allPods, err := cs.CoreV1().Pods(metav1.NamespaceAll).List(metav1.ListOptions{})
 		for i, node := range nodeList.Items {
-			cpuAllocatable, found := node.Status.Allocatable["cpu"]
+			currentCpuUsage, currentMemUsage := getCurrentPodUsageOnTheNode(node.Name, allPods.Items, podRequestedResource)
```
Comment (ravisantoshgudimetla) [inline vendor/k8s.io/kubernetes/test/e2e/scheduling/preemption.go]: This is the crux of the change, I am getting nodeObject again in every iteration.

**Pull request merged**